File: trunk/proj/CoreLibrary/src/com/jasonzqshen/familyaccounting/core/transaction/TransactionDataManagement.java
package com.jasonzqshen.familyaccounting.core.transaction;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.io.Writer;
import java.util.Hashtable;

import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import javax.xml.parsers.ParserConfigurationException;

import org.w3c.dom.Document;
import org.w3c.dom.Element;
import org.w3c.dom.Node;
import org.w3c.dom.NodeList;
import org.xml.sax.SAXException;

import com.jasonzqshen.familyaccounting.core.CoreDriver;
import com.jasonzqshen.familyaccounting.core.ManagementBase;
import com.jasonzqshen.familyaccounting.core.exception.*;
import com.jasonzqshen.familyaccounting.core.exception.runtime.*;
import com.jasonzqshen.familyaccounting.core.exception.format.*;
import com.jasonzqshen.familyaccounting.core.masterdata.MasterDataManagement;
import com.jasonzqshen.familyaccounting.core.utils.*;

public class TransactionDataManagement extends ManagementBase {

	public static final String TRANSACTION_DATA_FOLDER = "transaction_data";

	private final Hashtable<MonthIdentity, MonthLedger> _list;

	// private MonthLedger _openLedger; // current open ledger

	private final MasterDataManagement _masterDataMgmt;

	private final GLAccountBalanceCollection _glAccBalCol;

	/**
	 * constructor
	 * 
	 * @param coreDriver
	 */
	public TransactionDataManagement(CoreDriver coreDriver,
			MasterDataManagement masterDataMgmt) {
		super(coreDriver);
		_masterDataMgmt = masterDataMgmt;
		_list = new Hashtable<MonthIdentity, MonthLedger>();

		_glAccBalCol = new GLAccountBalanceCollection(_coreDriver,
				_masterDataMgmt);
	}

	/**
	 * get G/L account balance collection
	 * 
	 * @return
	 */
	public GLAccountBalanceCollection getAccBalCol() {
		return _glAccBalCol;
	}

	/**
	 * load data from file
	 * 
	 * @throws SystemException
	 *             bug
	 */
	public void initialize() throws TransactionDataFileFormatException {
		_coreDriver.logDebugInfo(this.getClass(), 75, String.format(
				"Starting month identity is %s", _coreDriver.getStartMonthId()
						.toString()), MessageType.INFO);

		MonthIdentity[] monthIdSet = _coreDriver.getAllMonthIds();

		// loop all the month ledger file to load transaction data
		try {
			for (MonthIdentity monthId : monthIdSet) {
				_coreDriver.logDebugInfo(
						this.getClass(),
						75,
						String.format("loading month identity is %s",
								monthId.toString()), MessageType.INFO);

				load(monthId);
			}
			// load current month identity
			// _openLedger = load(cureMonthId);
		} catch (NoTransactionDataFileException e) {
			_coreDriver.logDebugInfo(this.getClass(), 99, e.toString(),
					MessageType.ERROR);
			throw new TransactionDataFileFormatException("");
		}
	}

	/**
	 * load data from file based on month identity
	 * 
	 * @param monthId
	 * @throws TransactionDataFileFormatException
	 * @throws NoTransactionDataFileException
	 * @throws SystemException
	 * @throws MandatoryFieldIsMissing
	 */
	public MonthLedger load(MonthIdentity monthId)
			throws TransactionDataFileFormatException,
			NoTransactionDataFileException {
		_coreDriver.logDebugInfo(
				this.getClass(),
				102,
				String.format("Loading transaction data %s ...",
						monthId.toString()), MessageType.INFO);

		// get file path
		String filePath = generateFilePath(monthId);
		_coreDriver.logDebugInfo(this.getClass(), 109,
				String.format("Transaction data file: %s .", filePath),
				MessageType.INFO);

		// construct month ledger
		MonthLedger monthledger = new MonthLedger(monthId);
		_list.put(monthId, monthledger);

		File file = new File(filePath);
		// check if file exist
		if (!file.exists()) {
			return monthledger;
		}
		try {

			DocumentBuilderFactory docFactory = DocumentBuilderFactory
					.newInstance();
			DocumentBuilder builder = docFactory.newDocumentBuilder();
			Document doc = builder.parse(file);

			// get root element
			NodeList nodeList = doc.getChildNodes();
			Element rootElem = null;

			// get root element
			for (int i = 0; i < nodeList.getLength(); ++i) {
				Node child = nodeList.item(i);
				if (child instanceof Element) {
					Element elem = (Element) child;
					String nodeName = elem.getNodeName();
					if (nodeName.equals(TransDataUtils.XML_ROOT)) {
						rootElem = elem;
						break;
					}
				}
			}
			// no root element
			if (rootElem == null) {
				throw new TransactionDataFileFormatException(filePath);
			}

			// -------------------------------------------------------------------
			// parse all the documents
			nodeList = rootElem.getChildNodes();
			for (int i = 0; i < nodeList.getLength(); ++i) {
				Node child = nodeList.item(i);
				if (child instanceof Element) {
					Element elem = (Element) child;
					if (elem.getNodeName().equals(TransDataUtils.XML_DOCUMENT)) {

						HeadEntity head = HeadEntity.parse(_coreDriver,
								_masterDataMgmt, elem);
						_coreDriver
								.logDebugInfo(
										this.getClass(),
										172,
										String.format(
												"Document %s add to list during loading.",
												head.getDocIdentity()
														.toString()),
										MessageType.INFO);
						monthledger.add(head);

						// raise load document
						_coreDriver.getListenersManagement()
								.loadDoc(this, head);

						// raise reverse document
						if (head.IsReversed())
							_coreDriver.getListenersManagement()
									.reverseDoc(head);
					}
				}
			}
			// -----------------------------------------------------------------

			return monthledger;
		} catch (ParserConfigurationException e) {
			_coreDriver.logDebugInfo(this.getClass(), 170, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		} catch (SAXException e) {
			_coreDriver.logDebugInfo(this.getClass(), 170, e.toString(),
					MessageType.ERROR);
			throw new TransactionDataFileFormatException(filePath);
		} catch (IOException e) {
			_coreDriver.logDebugInfo(this.getClass(), 170, e.toString(),
					MessageType.ERROR);
			throw new NoTransactionDataFileException(filePath);
		} catch (TransactionDataFileFormatException e) {
			_coreDriver.logDebugInfo(this.getClass(), 170, e.toString(),
					MessageType.ERROR);
			throw new TransactionDataFileFormatException(filePath);
		}

	}

	/**
	 * generate transaction file path for the month identity
	 * 
	 * @param str
	 * @return
	 */
	private String generateFilePath(MonthIdentity monthId) {
		return String.format("%s/%s/%s.xml", _coreDriver.getRootPath(),
				TRANSACTION_DATA_FOLDER, monthId.toString());
	}

	/**
	 * save document
	 * 
	 * @param head
	 * @throws SaveClosedLedgerException
	 * @throws StorageException
	 */
	void saveDocument(HeadEntity head, boolean needStroe)
			throws SaveClosedLedgerException, StorageException {
		_coreDriver.logDebugInfo(this.getClass(), 231,
				"Call transaction to save the document", MessageType.INFO);

		if (head.isSaved()) {
			_coreDriver.logDebugInfo(this.getClass(), 235,
					"Document is saved, just to store the update on disk.",
					MessageType.INFO);
		} else {
			_coreDriver.logDebugInfo(this.getClass(), 239,
					"Document is never saved", MessageType.INFO);

			MonthIdentity monthId = head.getMonthId();
			MonthLedger ledger = this.getLedger(monthId);
			if (ledger == null) {
				_coreDriver.logDebugInfo(this.getClass(), 239,
						"Error in document month identity", MessageType.ERROR);
				throw new SaveClosedLedgerException();
			}

			// set document number
			DocumentNumber num = null;
			HeadEntity[] entities = ledger.getEntities();
			if (entities.length == 0) {
				try {
					num = new DocumentNumber("1000000001".toCharArray());
				} catch (IdentityTooLong e) {
					_coreDriver.logDebugInfo(this.getClass(), 239,
							e.toString(), MessageType.ERROR);
					throw new SystemException(e);
				} catch (IdentityNoData e) {
					_coreDriver.logDebugInfo(this.getClass(), 239,
							e.toString(), MessageType.ERROR);
					throw new SystemException(e);
				} catch (IdentityInvalidChar e) {
					_coreDriver.logDebugInfo(this.getClass(), 239,
							e.toString(), MessageType.ERROR);
					throw new SystemException(e);
				}
			} else {
				HeadEntity last = entities[entities.length - 1];
				num = last.getDocumentNumber().next();
			}
			_coreDriver.logDebugInfo(this.getClass(), 239,
					"Generate document number " + num.toString(),
					MessageType.INFO);
			head._docNumber = num;

			ledger.add(head);
		}

		if (needStroe) {
			store(head.getMonthId());
			_coreDriver.logDebugInfo(this.getClass(), 465,
					"Memory has been stored to disk", MessageType.INFO);
		} else {
			_coreDriver.logDebugInfo(this.getClass(), 465,
					"Memory has NOT been stored to disk", MessageType.INFO);
		}

		_coreDriver.logDebugInfo(this.getClass(), 278,
				"Call transaction to save the document successfully",
				MessageType.INFO);
	}

	/**
	 * store based on month identity
	 * 
	 * @param monthId
	 * @throws SystemException
	 */
	public void store(MonthIdentity monthId) throws StorageException {
		// store master data
		_coreDriver.logDebugInfo(this.getClass(), 293,
				String.format("Start storing %s to disk", monthId),
				MessageType.INFO);

		_coreDriver.logDebugInfo(this.getClass(), 297,
				"Store master data at first.", MessageType.INFO);
		MasterDataManagement manage = _coreDriver.getMasterDataManagement();
		manage.store();

		MonthLedger collection = this.getLedger(monthId);

		String filePath = this.generateFilePath(monthId);
		_coreDriver.logDebugInfo(this.getClass(), 297, "Generate file path: "
				+ filePath, MessageType.INFO);

		File file = new File(filePath);
		if (!file.exists()) {
			try {
				file.createNewFile();
			} catch (IOException e) {
				_coreDriver.logDebugInfo(this.getClass(), 304, e.toString(),
						MessageType.ERROR);
				throw new StorageException();
			}
		}
		String xdoc = collection.toXML();
		_coreDriver
				.logDebugInfo(this.getClass(), 297,
						"Parsed document collections to XML document",
						MessageType.INFO);

		Writer writer;
		try {
			writer = new BufferedWriter(new OutputStreamWriter(
					new FileOutputStream(file), XMLTransfer.default_charset));

			String header = null;
			Language lang = _coreDriver.getLanguage();
			if (lang == Language.Engilish) {
				header = Language.ENGLISH_XML_HEADER;
			} else if (lang == Language.SimpleChinese) {
				header = Language.SIMPLE_CHINESE_XML_HEADER;
			}
			writer.write(header, 0, header.length());
			writer.write(xdoc, 0, xdoc.length());
			writer.close();
		} catch (IOException e) {
			_coreDriver.logDebugInfo(this.getClass(), 316, e.toString(),
					MessageType.ERROR);
			throw new SystemException(e);
		}

		_coreDriver.logDebugInfo(this.getClass(), 335,
				"Save document collection successfully.", MessageType.INFO);
	}

	/**
	 * 
	 * @param docId
	 * @param msgs
	 * @return
	 * @throws SaveClosedLedgerException
	 * @throws ReverseOrgDocNotExistException
	 * @throws DocReservedException
	 */
	public void reverseDocument(DocumentIdentity docId)
			throws ReverseOrgDocNotExistException, DocReservedException {
		_coreDriver.logDebugInfo(this.getClass(), 348,
				"Start reversing document " + docId.toString(),
				MessageType.INFO);

		// ------------------------------------------
		// check document
		HeadEntity orgHead = this.getEntity(docId);
		if (orgHead == null) {
			_coreDriver.logDebugInfo(this.getClass(), 348, "No such document",
					MessageType.ERROR);
			throw new ReverseOrgDocNotExistException();
		}
		if (orgHead.IsReversed()) {
			_coreDriver.logDebugInfo(this.getClass(), 348,
					"Document has been reserved before", MessageType.ERROR);
			throw new DocReservedException();
		}

		try {
			_coreDriver.logDebugInfo(this.getClass(), 399,
					"Update reverse document information", MessageType.INFO);
			// update reverse information
			orgHead._isReversed = true;

			// raise event to update balance
			_coreDriver.getListenersManagement().reverseDoc(orgHead);

			_coreDriver.logDebugInfo(this.getClass(), 399,
					"Store memory to disk during reverse document",
					MessageType.INFO);

			// store to file system
			this.store(orgHead.getMonthId());
		} catch (StorageException e) {
			_coreDriver.logDebugInfo(this.getClass(), 452,
					"Dirty data does not store to file system",
					MessageType.WARNING);
		}

		String info = String.format("Document %s is reversed.",
				orgHead.getDocumentNumber());

		_coreDriver.logDebugInfo(this.getClass(), 416, info, MessageType.INFO);

	}

	/**
	 * get month ledger
	 * 
	 * @param monthId
	 * @return
	 */
	public MonthLedger getLedger(MonthIdentity monthId) {
		// get current calendar month
		MonthIdentity curMonthId = _coreDriver.getCurCalendarMonthId();

		// check whether ledger beyond the range
		if (_coreDriver.getStartMonthId().compareTo(monthId) > 0
				|| curMonthId.compareTo(monthId) < 0) {
			return null;
		}

		// create new month ledger
		if (!_list.containsKey(monthId)) {
			MonthLedger ledger = new MonthLedger(monthId);
			_list.put(monthId, ledger);
			return ledger;
		}

		return _list.get(monthId);
	}

	/**
	 * get documents
	 * 
	 * @param fiscalYear
	 * @param fiscalMonth
	 * @return
	 */
	public MonthLedger getLedger(int fiscalYear, int fiscalMonth) {
		MonthIdentity monthId;
		try {
			monthId = new MonthIdentity(fiscalYear, fiscalMonth);
		} catch (FiscalYearRangeException e) {
			return null;
		} catch (FiscalMonthRangeException e) {
			return null;
		}

		return getLedger(monthId);
	}

	/**
	 * get all month identities
	 * 
	 * @return
	 */
	public MonthIdentity[] getAllMonthIds() {
		return _coreDriver.getAllMonthIds();
	}

	/**
	 * get document based on the document identity
	 * 
	 * @param docId
	 * @return
	 */
	public HeadEntity getEntity(DocumentIdentity docId) {
		MonthLedger ledger = this.getLedger(docId._monthIdentity);
		if (ledger == null) {
			return null;
		}
		return ledger.getEntity(docId);
	}

	/**
	 * clear
	 */
	public void clear() {
		_list.clear();
	}

	@Override
	public void establishFiles() {
		// set up transaction data folder
		String transFolderPath = String.format("%s/%s",
				_coreDriver.getRootPath(), TRANSACTION_DATA_FOLDER);
		File transFolder = new File(transFolderPath);
		if (!transFolder.exists()) {
			_coreDriver
					.logDebugInfo(
							this.getClass(),
							125,
							"Transaction data root folder does not exist. Make directory.",
							MessageType.INFO);
			transFolder.mkdir();
		}

		MonthIdentity[] monthIds = _coreDriver.getAllMonthIds();
		for (MonthIdentity monthId : monthIds) {
			// create transaction file for each available month ledger
			MonthLedger ledger = new MonthLedger(monthId);
			_list.put(monthId, ledger);
			String xdoc = ledger.toXML();

			String filePath = generateFilePath(monthId);
			try {
				File file = new File(filePath);
				if (!file.exists()) {
					file.createNewFile();
				}
				BufferedWriter writer = new BufferedWriter(
						new OutputStreamWriter(new FileOutputStream(file),
								XMLTransfer.default_charset));
				writer.write(xdoc, 0, xdoc.length());
				writer.close();
			} catch (IOException e) {
				_coreDriver.logDebugInfo(this.getClass(), 530, e.toString(),
						MessageType.ERROR);
				throw new SystemException(e);
			}
		}
	}

}

File: branches/V1.0/proj/WinStore_FamilyAccounting/WinStore_FamilyAccountingCore/Transaction/TransactionDataManagement.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Xml.Linq;
using Windows.Storage;
using WinStore_FamilyAccountingCore.Exceptions;
using WinStore_FamilyAccountingCore.Exceptions.FormatExceptions;
using WinStore_FamilyAccountingCore.MasterData;
using WinStore_FamilyAccountingCore.Utilities;

namespace WinStore_FamilyAccountingCore.Transaction
{
    public class TransactionDataManagement : AbstractManagement
    {

        public static readonly String TRANSACTION_DATA_FOLDER = "transaction_data";

        private readonly Dictionary<MonthIdentity, MonthLedger> _list;

        // private MonthLedger _openLedger; // current open ledger

        private readonly MasterDataManagement _masterDataMgmt;

        private readonly GLAccountBalanceCollection _glAccBalCol;
        public GLAccountBalanceCollection AccountBalanceCol { get { return _glAccBalCol; } }

        /// <summary>
        /// construct
        /// </summary>
        /// <param name="coreDriver"></param>
        /// <param name="mdMgmt"></param>
        public TransactionDataManagement(CoreDriver coreDriver, MasterDataManagement mdMgmt)
            : base(coreDriver)
        {
            _masterDataMgmt = mdMgmt;
            _list = new Dictionary<MonthIdentity, MonthLedger>();

            _glAccBalCol = new GLAccountBalanceCollection(_coreDriver,
                    _masterDataMgmt);
        }

        /// <summary>
        /// 
        /// </summary>
        /// <returns></returns>
        /// <exception cref="TransactionDataFileFormatException"></exception>
        public override async Task InitializeAsync()
        {
            _coreDriver.logDebugInfo(this.GetType(), 75, String.Format(
                "Starting month identity is {0}", _coreDriver.StartMonthId
                        .ToString()), MessageType.INFO);

            MonthIdentity[] monthIdSet = _coreDriver.MonthIds;

            // loop all the month ledger file to load transaction data
            foreach (MonthIdentity monthId in monthIdSet)
            {
                _coreDriver.logDebugInfo(
                        this.GetType(),
                        75,
                        String.Format("loading month identity is {0}",
                                monthId.ToString()), MessageType.INFO);

                await Load(monthId);
            }
            // load current month identity
            // _openLedger = load(cureMonthId);

        }

        /// <summary>
        /// load month identity
        /// </summary>
        /// <param name="monthId"></param>
        /// <returns></returns>
        /// <exception cref="TransactionDataFileFormatException"></exception>
        public async Task<MonthLedger> Load(MonthIdentity monthId)
        {
            // construct month ledger
            MonthLedger monthledger = new MonthLedger(monthId);
            _list.Add(monthId, monthledger);

            _coreDriver.logDebugInfo(
                    this.GetType(),
                    102,
                    String.Format("Loading transaction data {0} ...",
                            monthId.ToString()), MessageType.INFO);

            // get file path
            StorageFile file = await generateFilePath(monthId, false);
            if (file == null)
            {// empty
                return monthledger;
            }
            _coreDriver.logDebugInfo(this.GetType(), 109,
                    String.Format("Transaction data file: {0} .", file.Path),
                    MessageType.INFO);
            try
            {
                string text = await FileIO.ReadTextAsync(file);
                XDocument xdoc = XDocument.Parse(text);

                XElement rootElem = xdoc.Element(TransDataUtils.XML_ROOT);
                // no root element
                if (rootElem == null)
                {
                    throw new TransactionDataFileFormatException(file.Path);
                }

                // -------------------------------------------------------------------
                // parse all the documents
                foreach (XElement elem in rootElem.Elements(TransDataUtils.XML_DOCUMENT))
                {
                    HeadEntity head = HeadEntity.Parse(_coreDriver,
                            _masterDataMgmt, elem);
                    _coreDriver
                            .logDebugInfo(
                                    this.GetType(),
                                    172,
                                    String.Format(
                                            "Document {0} add to list during loading.",
                                            head.DocIdentity
                                                    .ToString()),
                                    MessageType.INFO);
                    monthledger.Add(head);

                    // raise load document
                    _coreDriver.ListenerMgmt.LoadDoc(this, head);

                    // raise reverse document
                    if (head.IsReversed)
                        _coreDriver.ListenerMgmt.ReverseDoc(head);
                }
                // -----------------------------------------------------------------

                return monthledger;
            }
            catch (TransactionDataFileFormatException e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 170, e.Message,
                        MessageType.ERRO);
                throw new TransactionDataFileFormatException(file.Path);
            }

        }

        /// <summary>
        /// get file 
        /// </summary>
        /// <param name="monthId"></param>
        /// <returns></returns>
        /// <exception cref="SystemException"></exception>
        private async Task<StorageFile> generateFilePath(MonthIdentity monthId, bool createIfNotExist)
        {
            StorageFolder transactionFolder = await _coreDriver.RootFolder
                .GetFolderAsync(TRANSACTION_DATA_FOLDER);
            if (transactionFolder == null)
            {
                throw new SystemException(null); // bug
            }
            String fileName = String.Format("{0}.xml", monthId.ToString());
            
            if (createIfNotExist == false)
            {
                try
                {
                    StorageFile file = await transactionFolder.GetFileAsync(fileName);
                    return file;
                }
                catch (FileNotFoundException)
                {
                    return null;
                }
            }
            else
            {
                StorageFile file = await transactionFolder.CreateFileAsync(fileName
                        , CreationCollisionOption.OpenIfExists);
                return file;
            }
        }

        /// <summary>
        /// save document
        /// </summary>
        /// <param name="head"></param>
        /// <param name="needStroe"></param>
        /// <exception cref="SystemException"></exception>
        /// <exception cref="SaveClosedLedgerException"></exception>
        internal async Task saveDocumentAsync(HeadEntity head, bool needStroe)
        {
            _coreDriver.logDebugInfo(this.GetType(), 231,
                    "Call transaction to save the document", MessageType.INFO);

            if (head.IsSaved)
            {
                _coreDriver.logDebugInfo(this.GetType(), 235,
                        "Document is saved, just to store the update on disk.",
                        MessageType.INFO);
            }
            else
            {
                _coreDriver.logDebugInfo(this.GetType(), 239,
                        "Document is never saved", MessageType.INFO);

                MonthIdentity monthId = head.MonthID;
                MonthLedger ledger = this.GetLedger(monthId);
                if (ledger == null)
                {
                    _coreDriver.logDebugInfo(this.GetType(), 239,
                            "Error in document month identity", MessageType.ERRO);
                    throw new SaveClosedLedgerException();
                }

                // set document number
                DocumentNumber num;
                List<HeadEntity> entities = ledger.Entities;
                if (entities.Count == 0)
                {
                    try
                    {
                        num = new DocumentNumber("1000000001".ToCharArray());
                    }
                    catch (IdentityTooLong e)
                    {
                        _coreDriver.logDebugInfo(this.GetType(), 239,
                                e.ToString(), MessageType.ERRO);
                        throw new SystemException(e);
                    }
                    catch (IdentityNoData e)
                    {
                        _coreDriver.logDebugInfo(this.GetType(), 239,
                                e.ToString(), MessageType.ERRO);
                        throw new SystemException(e);
                    }
                    catch (IdentityInvalidChar e)
                    {
                        _coreDriver.logDebugInfo(this.GetType(), 239,
                                e.ToString(), MessageType.ERRO);
                        throw new SystemException(e);
                    }
                }
                else
                {
                    HeadEntity last = entities[entities.Count - 1];
                    num = last.DocNumber.Next();
                }
                _coreDriver.logDebugInfo(this.GetType(), 239,
                        "Generate document number " + num.ToString(),
                        MessageType.INFO);
                head._docNumber = num;

                ledger.Add(head);
            }

            if (needStroe)
            {
                await StoreAsync(head.MonthID);
                _coreDriver.logDebugInfo(this.GetType(), 465,
                        "Memory has been stored to disk", MessageType.INFO);
            }
            else
            {
                _coreDriver.logDebugInfo(this.GetType(), 465,
                        "Memory has NOT been stored to disk", MessageType.INFO);
            }

            _coreDriver.logDebugInfo(this.GetType(), 278,
                    "Call transaction to save the document successfully",
                    MessageType.INFO);
        }

        /// <summary>
        /// store month ledger to file system
        /// </summary>
        /// <param name="monthId"></param>
        /// <exception cref="SystemException"></exception>
        public async Task StoreAsync(MonthIdentity monthId)
        {
            // store master data
            _coreDriver.logDebugInfo(this.GetType(), 293,
                    String.Format("Start storing %s to disk", monthId),
                    MessageType.INFO);

            _coreDriver.logDebugInfo(this.GetType(), 297,
                    "Store master data at first.", MessageType.INFO);
            await this._masterDataMgmt.StoreAsync();

            MonthLedger collection = this.GetLedger(monthId);

            StorageFile file = await this.generateFilePath(monthId, true);
            _coreDriver.logDebugInfo(this.GetType(), 297, "Generate file path: "
                    + file.Path, MessageType.INFO);
            if (file == null)
            {
                StorageFolder transactionFolder = await _coreDriver.RootFolder
                .GetFolderAsync(TRANSACTION_DATA_FOLDER);
                if (transactionFolder == null)
                {
                    throw new SystemException(null); // bug
                }
                String fileName = String.Format("{0}.xml", monthId.ToString());
                file = await transactionFolder.CreateFileAsync(fileName
                    , CreationCollisionOption.OpenIfExists);
            }


            XDocument xdoc = collection.toXML();
            _coreDriver
                    .logDebugInfo(this.GetType(), 297,
                            "Parsed document collections to XML document",
                            MessageType.INFO);
            try
            {
                await FileIO.WriteTextAsync(file, xdoc.ToString());
            }
            catch (FileNotFoundException e)
            {
                _coreDriver.logDebugInfo(this.GetType(), 316, e.Message,
                        MessageType.ERRO);
                throw new SystemException(e);
            }

            _coreDriver.logDebugInfo(this.GetType(), 335,
                    "Save document collection successfully.", MessageType.INFO);
        }

        /// <summary>
        /// store month identity
        /// </summary>
        /// <param name="monthId"></param>
        public async void Store(MonthIdentity monthId)
        {
            await StoreAsync(monthId);
        }

        /// <summary>
        /// Get Ledger
        /// </summary>
        /// <param name="monthId"></param>
        /// <returns></returns>
        public MonthLedger GetLedger(MonthIdentity monthId)
        {
            // get current calendar month
            MonthIdentity curMonthId = _coreDriver.CurCalendarMonthId;

            // check whether ledger beyond the range
            if (_coreDriver.StartMonthId.CompareTo(monthId) > 0
                    || curMonthId.CompareTo(monthId) < 0)
            {
                return null;
            }

            // create new month ledger
            MonthLedger monthLedger;
            if (!_list.TryGetValue(monthId, out monthLedger))
            {
                monthLedger = new MonthLedger(monthId);
                _list.Add(monthId, monthLedger);
            }

            return monthLedger;
        }


        /// <summary>
        /// get ledger
        /// </summary>
        /// <param name="fiscalYear"></param>
        /// <param name="fiscalMonth"></param>
        /// <returns></returns>
        public MonthLedger GetLedger(int fiscalYear, int fiscalMonth)
        {
            MonthIdentity monthId;
            try
            {
                monthId = new MonthIdentity(fiscalYear, fiscalMonth);
            }
            catch (FiscalYearRangeException)
            {
                return null;
            }
            catch (FiscalMonthRangeException)
            {
                return null;
            }

            return GetLedger(monthId);
        }

        /// <summary>
        /// Reverse Document
        /// </summary>
        /// <param name="docId"></param>
        /// <exception cref="SystemException"></exception>
        /// <exception cref="ReverseOrgDocNotExistException"></exception>
        /// <exception cref="DocReservedException"></exception>
        public void ReverseDocument(DocumentIdentity docId)
        {
            _coreDriver.logDebugInfo(this.GetType(), 348,
                    "Start reversing document " + docId.ToString(),
                    MessageType.INFO);

            // ------------------------------------------
            // check document
            HeadEntity orgHead = this.GetEntity(docId);
            if (orgHead == null)
            {
                _coreDriver.logDebugInfo(this.GetType(), 348, "No such document",
                        MessageType.ERRO);
                throw new ReverseOrgDocNotExistException();
            }
            if (orgHead.IsReversed)
            {
                _coreDriver.logDebugInfo(this.GetType(), 348,
                        "Document has been reserved before", MessageType.ERRO);
                throw new DocReservedException();
            }

            _coreDriver.logDebugInfo(this.GetType(), 399,
                    "Update reverse document information", MessageType.INFO);
            // update reverse information
            orgHead._isReversed = true;

            // raise event to update balance
            _coreDriver.ListenerMgmt.ReverseDoc(orgHead);

            _coreDriver.logDebugInfo(this.GetType(), 399,
                    "Store memory to disk during reverse document",
                    MessageType.INFO);

            // store to file system
            this.Store(orgHead.MonthID);


            String info = String.Format("Document {0} is reversed.",
                    orgHead.DocNumber);

            _coreDriver.logDebugInfo(this.GetType(), 416, info, MessageType.INFO);

        }


        /// <summary>
        /// get entity
        /// </summary>
        /// <param name="docId"></param>
        /// <returns></returns>
        public HeadEntity GetEntity(DocumentIdentity docId)
        {
            MonthLedger ledger = this.GetLedger(docId._monthIdentity);
            if (ledger == null)
            {
                return null;
            }
            return ledger.GetEntity(docId);
        }

        /// <summary>
        /// clear
        /// </summary>
        public override void Clear()
        {
            _list.Clear();
        }

        /// <summary>
        /// establish files
        /// </summary>
        /// <returns></returns>
        public override async Task EstablishFilesAsync()
        {
            // set up transaction data folder
            await _coreDriver.RootFolder.CreateFolderAsync(TRANSACTION_DATA_FOLDER
                , CreationCollisionOption.OpenIfExists);
            _coreDriver
                    .logDebugInfo(
                            this.GetType(),
                            125,
                            "Create transaction data root folder.",
                            MessageType.INFO);

            foreach (MonthIdentity monthId in _coreDriver.MonthIds)
            {
                // create transaction file for each available month ledger
                MonthLedger ledger = new MonthLedger(monthId);
                _list.Add(monthId, ledger);
                XDocument xdoc = ledger.toXML();

                StorageFile file = await generateFilePath(monthId, true);
                try
                {
                    await FileIO.WriteTextAsync(file, xdoc.ToString());
                }
                catch (FileNotFoundException e)
                {
                    _coreDriver.logDebugInfo(this.GetType(), 530, e.Message,
                            MessageType.ERRO);
                    throw new SystemException(e);
                }
            }
        }

        public override bool NeedInit
        {
            get { return true; }
        }

        public override bool NeedEstablishFile
        {
            get { return true; }
        }
    }
}
